Change the e2e selector for unchecking all business features

tammullen opened the pull request:
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

There have been a large number of failures in CI for `can setup shipping`.

After debugging the majority of these failures appear to be because smart shipping zones haven't been setup for a `US` store.
The most likely reason for this is `Add recommended business features to my site` has not been unchecked due to test flakiness and therefore WC Shipping installed which would mean no smart shipping zones would be setup.

To try prevent this from happening this PR updates the selector used to uncheck the `Add recommended business features to my site` checkbox to [match the selector used in `admin-e2e-tests`](https://github.com/woocommerce/woocommerce/blob/a0768a086ca73747957273457fc6151be3e19b4c/packages/js/admin-e2e-tests/src/sections/onboarding/BusinessSection.ts#L91)

It also adds some additional beforeEach setup hooks for the `can setup shipping` test to uncheck the `Add recommended business features to my site` checkbox if it is checked.

### How to test the changes in this Pull Request:

1. [Run e2e tests locally](https://github.com/woocommerce/woocommerce/tree/trunk/plugins/woocommerce/tests/e2e-pw#introduction)
2. [Check e2e CI results](https://woocommerce.github.io/woocommerce-test-reports/pr/34685/e2e/)

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [x] Have you written new tests for your changes, as applicable?
-   [x] Have you successfully run tests with your changes locally?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm changelog add --filter=<project>`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

Commit:
Change the e2e selector for unchecking all business features

Commit:
Update e2e uncheck business features in test setup

adrianduffell commented on plugins/woocommerce/tests/e2e-pw/tests/activate-and-setup/complete-onboarding-wizard.spec.js:
```diff
@@ -227,8 +227,8 @@ test.describe(
 			// Uncheck all business features
-			if ( page.isChecked( '#inspector-checkbox-control-1' ) ) {
-				await page.click( '#inspector-checkbox-control-1' );
+			if ( page.isChecked( '.components-checkbox-control__input' ) ) {
+				await page.click( '.components-checkbox-control__input' );
```
I think this will match the first checkbox on the page. This should work ok for now, but might be flaky in the future if another checkbox is added elsewhere on the page before it.

There doesn't seem like a way we can match the checkbox precisely at the moment. It's missing an associated `label` so we can't do a text match either. That's probably an accessibility concern as well.

I'll create a follow-up issue to make this checkbox easier to match.

alopezari replied:
Thanks @adrianduffell!

The good practices suggest using the custom [`data-*` attributes](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/data-*) to access elements for test automation, such as `data-test` or `data-test-id`, for example, as requested in the following [Selenium PR](https://github.com/SeleniumHQ/selenium-ide/issues/435) or in the [good practices recommended by Cypress](https://docs.cypress.io/guides/references/best-practices#Selecting-Elements).

The reason for this is that even if we use a specific `className` or `id` of the element in our tests, these might change as part of a refactor and the test would need to be updated. On the other hand, if we use custom `data-*` locators in our tests, these locators will be specifically created for this purpose, so we would decouple the automated tests from the front-end implementation details.

It would be great if we could start adding these `data-*` custom attributes to elements from now on 🔥

adrianduffell replied:
@alopezari I'd agree, that would be great 🔥 

Unfortunately the underlying WordPress/Gutenberg [CheckboxControl component](https://github.com/WordPress/gutenberg/tree/a3c753d28ddb0972311e2da0136277626bdf832e/packages/components/src/checkbox-control) doesn't allow us to configure it with data attributes, but maybe we could try to move forward with a PR in their repo to bring in support for it 🤔

alopezari replied:
Oh, I didn't know about that, thanks for letting me know @adrianduffell!

> but maybe we could try to move forward with a PR in their repo to bring in support for it

That could be quite interesting. And if that PR becomes successful, it would leave the door open to do the same with other components 😲

tammullen replied:
> I think this will match the first checkbox on the page. This should work ok for now, but might be flaky in the future if another checkbox is added elsewhere on the page before it.
> 
> There doesn't seem like a way we can match the checkbox precisely at the moment. It's missing an associated label so we can't do a text match either. That's probably an accessibility concern as well.

Yes agreed, I think switching to the class for selector isn't ideal but is the only way to improve the selector for now.

> I'll create a follow-up issue to make this checkbox easier to match.

Thank you!

alopezari commented:
Thanks for creating this PR @tammullen!

I ran it locally and works on most cases in my local but there is one scenario where the `Store owner can go through setup Task List › can setup shipping` test fails: 

![image](https://user-images.githubusercontent.com/45979455/190344943-88080397-f70f-4860-b850-967020df2f70.png)
[https://d.pr/i/h1Dvcl](https://d.pr/i/h1Dvcl)

This failure is consistent and repeatable, and seems it only occurs when I run the tests in headless mode after setting up a fresh environment (making sure I deleted the previous one).

For example:
- It passes when I ran the tests in headed mode.
- It passes when I ran the tests in headless mode after I had run it previously headed.
- It fails when I run the tests in headless mode directly after setting up a new environment.

It would be great if you could try to reproduce the failing scenario above @tammullen, just to confirm if it's actually failing or there is some gap in my local environment. Seems tests run sequentially when running a single spec file so, if it turns out this is an actual issue, could it be a race condition?

PS: I've found this video as part of the test execution, but I don't know if the page gets stuck or is the recorder the one that gets stuck 😅 :
[d3f8f3c67457c4d30265ab34fd6e900dad42858a.webm](https://user-images.githubusercontent.com/45979455/190375607-7b40584f-ac2d-4080-9b6c-9f7f87d66fa5.webm)

Commit:
Add theme selection to before hook for shipping task

tammullen commented:
> It fails when I run the tests in headless mode directly after setting up a new environment.

Thanks @alopezari for catching this!

I was able to reproduce this. 
I had added extra steps to the before hooks to make sure the `Add recommended business features to my site` is unchecked. Both the original and my extra steps didn't actually exit the OBW so when trying to access  `admin.php?page=wc-admin` the re-direction kicked in and stopped it reaching the task list, not sure why it only causes failures on fresh site and not for the other ways or in CI.

To improve this I have added an extra step to exit the OBW on theme selection.

tammullen committed:
Added changelog

alopezari commented:
> To improve this I have added an extra step to exit the OBW on theme selection.

Thanks, @tammullen! It works great now, I'm approving this PR.
